crates/bevy_ecs/src/storage/sorted_small_vec.rs

```diff
@@ -0,0 +1,259 @@
+        let mut i = 0;
+        let mut j = 0;
+        while i < self.len() && j < other.len() {
+            match self.0[i].cmp(&other.0[j]) {
+                Ordering::Less => i += 1,
+                Ordering::Greater => {
+                    self.0.insert(i, other.0[j]);
+                    j += 1;
+                }
+                Ordering::Equal => {
+                    i += 1;
+                    j += 1;
+                }
+            }
+        }
+        while j < other.len() {
+            self.0.push(other.0[j]);
+            j += 1;
+        }
+    }
+
+    /// Returns the elements that are in both `self` and `other`.
+    pub(crate) fn intersection<'a>(&'a self, other: &'a Self) -> Intersection<'a, N> {
+        Intersection {
+            this: self,
+            other,
+            i: 0,
+            j: 0,
+        }
+    }
+
+    /// Return the elements that are in `self` but not in `other`.
+    pub(crate) fn difference<'a>(&'a self, other: &'a Self) -> Difference<'a, N> {
+        Difference {
+            this: self,
+            other,
+            i: 0,
+            j: 0,
+        }
+    }
+
+    /// Returns true if the two vectors have no common elements.
+    pub(crate) fn is_disjoint(&self, other: &Self) -> bool {
+        self.intersection(other).next().is_none()
+    }
+
+    /// Returns true if all the elements in `self` are also in `other`.
+    pub(crate) fn is_subset(&self, other: &Self) -> bool {
+        self.difference(other).next().is_none()
+    }
+}
+
+impl<const N: usize> Extend<usize> for SortedSmallVec<N> {
+    fn extend<T: IntoIterator<Item = usize>>(&mut self, other: T) {
+        let mut i = 0;
+        let mut other_iter = other.into_iter();
+        let mut other_val = other_iter.next();
+        while i < self.len() && other_val.is_some() {
+            let val_j = other_val.unwrap();
+            match self.0[i].cmp(&val_j) {
+                Ordering::Less => {
+                    i += 1;
+                }
+                Ordering::Greater => {
+                    self.0.insert(i, val_j);
+                    other_val = other_iter.next();
+                }
+                Ordering::Equal => {
+                    i += 1;
+                    other_val = other_iter.next();
+                }
+            }
+        }
+        while let Some(val_j) = other_val {
+            self.0.push(val_j);
+            other_val = other_iter.next();
+        }
+    }
+}
+
+/// Intersection between `this` and `other` sorted vectors.
+pub(crate) struct Intersection<'a, const N: usize> {
+    this: &'a SortedSmallVec<N>,
+    other: &'a SortedSmallVec<N>,
+    i: usize,
+    j: usize,
+}
+
+impl<'a, const N: usize> Iterator for Intersection<'a, N> {
+    type Item = usize;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut res = None;
+        while self.i < self.this.len() && self.j < self.other.len() {
+            if (self.i == 0 || self.this.0[self.i] != self.this.0[self.i - 1])
```

Comment:
I don't understand this expression. How can `[self.i] == [self.i - 1]`? We should never have the same value twice.

(If we change anything here, we should mirror it with `Difference`)

Reply:
I think that this accounted for possible duplicates, but only for self for some reason. Since we're already assuming that SortedVecSets are in fact 'sorted', I don't think that it's too much to ask for zero duplicates either. I've documented it.